Render dashed lines for all railways under construction in the electrification layer

JonathanBittner opened the pull request:
This PR fixes issue #47 where certain lines under construction show as completed lines in the electrification layer. I also addressed as a quick fix #42 but it is not as comprehensive as PR #57.

This wasn't a straightforward fix and required a bit of tweaking. We may also want to address the color scheme: the dashed pattern of grey and color for lines under construction does not render well on thinly weighted light rail lines.

Some of these issues were caused by poor or ambiguous tagging where there is a mix of regular (electrified,voltage,frequency) and `construction:` tags.  While some of these issues can be corrected in the project code, we should not attempt to become an AI bot and "do what the user meant." However, I will take advantage of certain fall throughs if they produce the desired result. An improper fall through in `railway_electrification_state()` [sql/functions.sql](../blob/master/sql/functions.sql) triggered several of the issues.  

Here are some of the behaviors I have documented:
1. [**OK**]  extant line, electrification upgrades under construction (`railway=rail`; `construction:electrification=rail/contact_line/4th_rail` => renders dashed line alternating between black and the assigned electrification color; example: [Caltrain San Francisco to San Jose](https://www.openrailwaymap.org/?style=electrified&lat=37.469246737714116&lon=-122.20658183097838&zoom=16)
2. [**OK**] line under construction, properly tagged construction:electrified/construction:voltage/construction:frequency => renders dashed grey + assigned electrification color; example: [here](https://www.openrailwaymap.org/?style=electrified&lat=45.20648803427833&lon=6.758286952972412&zoom=15) 
3. [**BAD!**] non-electrified line under construction, construction:electrified=no => renders as completed line/ambiguous electrification (solid grey); example: [NJ Transit Lackawanna Cut Off Expansion](https://www.openrailwaymap.org/?style=electrified&lat=40.95290460015215&lon=-74.70175266265869&zoom=14);  Desired behavior: render as dashed black.
4. [**BAD!**] electrified line under construction, electrified/voltage/frequency specified without `construction:` prefix => renders as a solid line in the as if the line were completed; example: the [Finland case](https://www.openrailwaymap.org/?style=electrified&lat=60.22936357183169&lon=24.91794168949127&zoom=17) in   Desired behavior: render as broken line with electrification colors (anticipating user needs) OR render as a dashed line without colorization indicating line is under construction. This could be corrected by fixing the tagging but I imagine we'll see many other ways with this style of tagging.  The example here was taken from OpenRailwayMap/OpenRailwayMap#718 
5.  [**BAD!**] line under construction, ambiguously tagged (e.g. `electrified=yes`) => renders as solid grey; example [here](https://www.openrailwaymap.org/?style=electrified&lat=19.39290029041924&lon=-99.19784590601921&zoom=19) We don't necessarily have a desired behavior for this, but the line should be dashed  In this PR I propose interleaving the current grey fill color with an additional `darkgrey` fill color to mix grey/darkgrey
6. around [Greenboro light rail station](https://www.openrailwaymap.org/?style=electrified&lat=45.35830104572409&lon=-75.65703213214873&zoom=17) renders solid black to the north of the station where it is fully tagged (`construction:railway=light_rail`; `construction=light_rail`; `railway=construction`)  but is also tagged `electrified=no`.  The line just south of the station has no electrification tagging and renders solid grey. Both lines should be dashed.


All 6 examples now render properly.

epimorphic commented:
I'm not sure about labeling example 2 as "OK" -- it seems to suggest the line already exists, just with incomplete tagging of the present electrification status. Having only the dashed line for the future electrification, without the background gray "no information" line, seems better to me. (The white casing around the fill lines seems excessive for half-transparent dashed lines though, so [I've been playing around with opacity for casing](https://epimorphic.github.io/ORM-dash-mockup/opacity.html).)

Also the PR as it stands turns Switzerland (and I'd imagine many of its neighbors) into a sea of yellow, something that also happened in the course of the US/Japan PR. (Really, all the more reason to finally do something about that one.)

JonathanBittner commented:
I agree with @epimorphic: going to a white background on new lines under construction is a good idea.  The gauge layer already does this.  By doing this, i can likely simplify both the SQL and CSS.  Give me a few more days...

JonathanBittner commented:
I started this one anew.  It will no longer attempt to correct issue 42, let PR #57 handle that.
I tried to simplify things but in the end things got a little more complicated.

I changed how the layers handle construction: lines under construction are handled in the railway_fill_layer and are set to dashed.  Color is assigned based on electrification/no electrification/unknown electrification  New lines under construction will no longer be handled by the electrification_future layer.  This layer will exclusively handle existing lines that are being upgraded to electrified (or are proposed to be upgraded).

While the CSS has become more simplified compared to my last submission, the SQL became a bit more complicated.  New states were added to railway_electrification_state() to remove ambiguites.  The states are "present," "no," null (undefined/ambiguous), "construction_now" (new, indicating a line under construction), "construction_future" (new, indicating a line being upgraded), and "proposed_future" (renamed from "proposed" for consistency.  There is no "proposed_now" state as the previous version did not render proposed new lines at all.

In addition, I had to make some more changes to the SQL to ensure construction was rendered in oblique fonts and certain pass-through cases were handled accordingly and the voltages and frequencies were captured correctly.  What do I mean by this?  For lines being built that have electrification included, there is quite a bit of mistagging in the database. The present code handled this reasonably well: rather than cause more breakage and disruption to the map, I added some additional handling to continue to capture these values. For instance a light rail line is under construction. It was tagged railway=construction, construction=light_rail, and construction:railway=light_rail.  On the electrification side, it was tagged: electrified=contact_line, voltage=750, and frequency=0.  It would be more correct to tag construction:electrified=contact_line, construction:voltage=750, and construction:frequency=0, but as I mentioned there are many lines tagged this way and we should be able to handle them

DerDakon commented on electrification.mss:
```diff
@@ -221,7 +214,7 @@
     }
 
-    [frequency!=null][frequency!=0][voltage>=15000][voltage<25000] {
+    [frequency=null][frequency!=0][voltage>=15000][voltage<25000] {
```
This change looks suspicious.

JonathanBittner replied:
It looks highly suspicious and downright wrong.  There is a bug covered in #42 / PR #57 that I may tried to briefly address but something may have gone wrong with the merging

The original code on line 224 is `[frequency=null][frequency!=0][voltage>=25000] ` which does nothing.  The code was supposed to assign a color to AC lines >25000 such as the mine-to-power plant lines in the SW United States.  

So the code should be `[frequency!=null][frequency!=0][voltage>=25000]` or maybe even '[frequency>0][voltage>=25000]' and I tried to briefly address this in my PR.  Perhaps since I might have merged the proposed changes from PR #57 and then derived changes, something is messed up here.

JonathanBittner replied:
Ok, the current code in the repo for electrification.mss on lines 212 and 224 is incorrect.  My submission doesn't change these lines even though they are wrong.  Essentially any AC line between 15 kV and 25 kV (like 20 kV in Japan) or any AC line above 25 kV is not rendered with a color right now.  While I haven't changed the code here, I was indeed working from another branch and the change highlighting makes it appear I've substituted bad code for good code when it's more like I'm leaving bad code but had a sample of good code that I'm letting PR #57 work out.

I could do a simple PR that addresses Issue #42 but is not as extensive as PR #57. PR #57 fixes these issues but also tries to add separate colors for major lines that are visible at zoom level 2 but aren't included in the existing color scheme, namely Japanese electrification (20 kV/50 Hz and 20 kV/60 Hz) and Metro-North/Amtrak 12.5 kV/60 Hz in New York and Connecticut.  I could then rebase after that PR is accepted.

DerDakon commented on electrification.mss:
```diff
@@ -233,7 +226,7 @@
     }
 
-    [frequency!=null][frequency!=0][voltage>25000] {
+    [frequency=null][frequency!=0][voltage>=25000] {
```
This change looks suspicious.

DerDakon commented:
I think it's time to update this as at least #57 has already been merged. Maybe we can even get an example rendering for a place where this is currently be wrong and then fixed by this commit?
